docs: Master the Shell workshop (#2)
lasselundstenjensen/itu-bds-sdse: lessons/03-golang-introduction/Z-(Optional)-Master-The-Shell-Workshop.md

---

# The shell as a tool

This is a small overview/exercise sheet on basic shell commands that will help you interact with files and programs in a graceful manner. The shell is a valuable tool and can in most instances be the fastest and most optimal way to work. People who are not used to working in the shell may see it purely as a tool to compile and run code but out of the box it comes with many powerful tools that let you skip clunky graphical user interfaces entirely.

</br>
</br>

## Which shell to use?
For this overview I will assume we are using either bash or zsh (all things covered in this guide should work the same in both out of the box)

## A great start

Most of this guide will be based on the ["missing-semester" - From MIT](https://missing.csail.mit.edu/2020/course-shell/)

- Introduction and basics [here](https://missing.csail.mit.edu/2020/course-shell/)
- Shell scripting [here](https://missing.csail.mit.edu/2020/shell-tools/)
- Command-line environment [here](https://missing.csail.mit.edu/2020/command-line/)

We will mostly cover basics and a small amount of shell scripting but feel free to read through these 3 ressources in depth.

# Lets stretch our legs

> One of the most important skills to learn is how to get get around the shell and interact with files. Without the ability to move, read, move or delete files we can't really do a whole lot :)

Reviewer: ```suggestion
One of the most important skills to learn is how to get get around the shell and interact with files. Without the ability to move, read, write or delete files we can't really do a whole lot :)
```

Explaining the symbols:

> user:<something>: user is a standin for your own username the : and then some word is our current directory name  

Reviewer: ```suggestion
user:<something>: user is a standin for your own username, while the : and then some word is our current directory name.
```
Though it might be different from WSL, MacOS, etc...

> \$: Anytime you see a \$ it purely marks the start of a shell line this might look different in your shell/terminal if you want to run a command feel free to copy it from the code blocks but do not take the \$ with you it will not run if you do. 

Reviewer: ```suggestion
\$: Anytime you see a \$, it marks the start of a shell line. This might look different in your shell. If you want to run a command, feel free to copy it from the code blocks, but do not take the \$ with you, as it will not run if you do. 
```

Writing `shell/terminal` makes the terms seem very interchangeable :')

\>: The carrot symbol (>) will be used to show the output of a given command

### cd (change directory)

> One of the most used commands will be `cd` which will change your directory to whatever directory you pass it. If left blank it will move you to (~/$HOME) which is your home directory.

Reviewer: ```suggestion
One of the most used commands will be `cd` which will change your directory to whatever directory you pass it. If left blank, it will move you to `~` or `$HOME`, which is your home directory.
```

```bash
user:current$ cd 
> user:~$ 
```

```bash
user:current$ cd cool_directory
> user:cool_directory$
```

### ls (list) 

> To be able to move around you must know what is around you the ls command is useful here as it lists files and directories.

Reviewer: ```suggestion
To be able to move around you must know what is around you. The `ls` command is useful here, as it lists files and directories within your current directory.
```

```
Notice how we get the exact same output as when we were in current and ran `ls`

### cat (con(cat)enate)
This is mostly used for reading files but as the name suggests it can also be used to concatenate files together

```bash
user:current$ cat cool_file.md
> # this is a super cool file
it even spans multiple lines
```

### echo (command tool used for displaying lines of text or string)
This will work somewhat like a print in python but its a little more powerful since it can be chained with other shell commands and options to achieve broader goals.

```bash
user:~$ echo hello
> hello
```

We could as an example capture the output of echo in a file (note this file doesn't even have to exist echo will create it for us)

```bash
user:~$ echo hello > file_name
> 
user:~$ cat file_name
> hello
user:~$ echo hello >> file_name
> hello
hello
```

Notice the carrot (>) behind `echo hello` this means "output to x" in this case we are telling it to output to the file named `file_name`. The double carrot (>>) means append so add to the end of.

### pwd (print working directory)
As the name suggests this command will print your current working directory aka. where you are

> Note this command will print the full path and thus may look slightly different on Linux and MacOS as the home folder is stored in different locations. If you are Linux `/User` will be `/home`

Reviewer: ```suggestion
Note this command will print the full path and thus may look slightly different on Linux and MacOS as the home folder is stored in different locations. If you are on Linux, `/Users` will be `/home`.
```

Reading longer files using cat is kind of annoying so instead we have another tool that we can use. If you want to know more then you can try `man less` 

```bash
user:current$ less cool_file.md
>
```
This will open a page allowing you to scroll through a file and even search for things. To exit use q

### tail 
This will give you the tail of a file aka the end. You can specify how much you want using flags and can even watch a file using this command. By default it will give you the whole file use `-n` with a number to specify how many lines e.g. `-n1` for just the last line

```bash
user:current$ tail -n1 cool_file.md
> it even spans multiple lines
```

You can watch the file using the -f flag. If somebody appended to the cool_file.md it would display under the cursor
```bash
user:current$ tail -f append_to_me
> 
```
WARNING: This will take control of your shell. If you want out your can always call a SIGKILL using ctrl+C

### find 
This command will let your find or search for files using a pattern. We will not cover what a pattern is here but a simple google search or even the man page may give you a good idea of how to create one that suits your case

> #TODO: FINISH

Reviewer: Left an unfinished TODO ^^"

Author: Resolved via manual commit

## Makefiles: A useful usecase for shell commands (and bash scripting)
`Makefiles` are useful in projects where you end up doing the same things a lot, like compiling with specific flags or running tests before compiling or other such examples. This could even be used in the project you will embark on later in this course! A `Makefile` in essence, is just a list of command names and a list of shell commands to run when given that command name (you can even pass arguments to the `Makefile` command ). A `Makefile` can be made by just creating a file called `Makefile` in your project directory. To run a command defined in your file simply run the below command:

```bash
user:current$ make <command>
> 
```
Where `<command>` is a command defined in your `Makefile`. If you wanted to run generate from the below example you could run the command `make generate`. By default `make` will run the first command defined by a name. 

Heres an example of a `Makefile` from a golang project 

```makefile
generate:
	echo "Generating..."
	cd src/tailwind && npm run build
	go run main.go generate 
```

> This `Makefile` will print "Generating..." to the terminal then run npm run build inside a folder named src/tailwind then after that run a go program from the main.go file.

Reviewer: ```suggestion
This `Makefile` will print "Generating..." to the terminal, then run `npm run build` inside a folder named `src/tailwind. After that, it will run a go program from the `main.go` file.
```
